File: Turquoise/UI/Common/TQTextField.m

#import "TQTextField.h"

@interface TQTextField ()<UITextFieldDelegate>
@end

@implementation TQTextField {
  NSMutableString *_password;
}

static NSString *const kAsteriskString = @"****************************************************************";
static const NSUInteger kMaxPasswordLength = 64;

+ (NSString *)hiddenStringForString:(NSString *)string {
  if (!string) {
    return nil;
  }

  NSUInteger length = MIN(kMaxPasswordLength, string.length);
  return [kAsteriskString substringToIndex:length];
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
  self = [super initWithCoder:aDecoder];
  if (self) {
    self.delegate = self;
    self.autocorrectionType = UITextAutocorrectionTypeNo;
    [self addTarget:self
               action:@selector(textUpdated)
      forControlEvents:UIControlEventEditingChanged];
  }
  return self;
}

- (void)setIsPassword:(BOOL)isPassword {
  if (isPassword != _isPassword) {
    _isPassword = isPassword;
    self.text = @"";
    _password = [NSMutableString string];
  }
}

- (NSString *)password {
  return [_password copy];
}

- (void)textUpdated {
  if (_isPassword) {
    self.text = [[self class] hiddenStringForString:self.text];
  }
}

#pragma mark - UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField
  shouldChangeCharactersInRange:(NSRange)range
              replacementString:(NSString *)string {
  if (_isPassword) {
    [_password replaceCharactersInRange:range withString:string];
  }
  return YES;
}

#pragma mark - UITextField Overrides

- (CGRect)textRectForBounds:(CGRect)bounds {
  return CGRectInset(bounds , 10, 0);
}
- (CGRect)editingRectForBounds:(CGRect)bounds {
  return CGRectInset(bounds , 10, 0);
}

@end

File: Turquoise/UI/Common/TQTextField.swift
import Foundation
import UIKit

class TQTextField : UITextField, UITextFieldDelegate {
  static let kMaxPasswordLength = 64

  var isPassword: Bool = false {
    didSet {
      text = ""
      password = isPassword ? "" : nil
    }
  }

  private(set) var password: String? = ""

  override var text: String? {
    get {
      return super.text
    }
    set {
      var newText = newValue
      if isPassword {
        if password?.count != newValue?.count {
          // this only happens when the text is modified programmatically.
          // when change is done through the UI, we already have the correct password here.
          password = newValue
        }
        newText = type(of: self).hiddenString(password)
      } else {
        password = ""
      }
      super.text = newText
    }
  }

  static func hiddenString(_ string: String?) -> String? {
    if let theString = string {
      let count = min(kMaxPasswordLength, theString.count)
      return String.init(repeating: "*", count: count)
    } else {
      return nil
    }
  }

  required init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    password = ""
    delegate = self
    autocorrectionType = UITextAutocorrectionType.no
    addTarget(self, action: #selector(textUpdated), for: UIControlEvents.editingChanged)
  }

  func textUpdated() {
    if isPassword {
      text = password
    }
  }

  // MARK: - UITextFieldDelegate

  func textField(_ textField: UITextField,
                 shouldChangeCharactersIn range: NSRange,
                 replacementString string: String) -> Bool {
    if isPassword {
      if let theRange = Range(range, in:text!) {
        password?.replaceSubrange(theRange, with: string)
      }
    }
    return true
  }

  func textFieldShouldClear(_ textField: UITextField) -> Bool {
    // text field will be cleared, clear password as well.
    password = ""
    return true
  }

  // MARK: - UITextField Overrides

  override func textRect(forBounds bounds: CGRect) -> CGRect {
    return bounds.insetBy(dx: 10, dy: 0)
  }

  override func editingRect(forBounds bounds: CGRect) -> CGRect {
    return bounds.insetBy(dx: 10, dy: 0)
  }
}
